### ISSUE TITLE
Concurrency problem when ranging through .Site.Menus

### ISSUE THREAD

**Issue by jmooring** [gohugoio/hugo#7594]
### What version of Hugo are you using (`hugo version`)?

<pre>
$ hugo version
Hugo Static Site Generator v0.75.0-DEV/extended linux/amd64 BuildDate: unknown
</pre>

### Does this issue reproduce with the latest release?

Yes

### Summary

When ranging through `.Site.Menus.foo.ByWeight` with `HUGO_NUMWORKERMULTIPLIER` greater than 1, the rendered result is intermittently incorrect, with items in the wrong order, or duplicate items. This does not occur when `HUGO_NUMWORKERMULTIPLIER` equals 1.

The problem occurs when running either `hugo` or `hugo server`.

See [Forum Topic #27710](https://discourse.gohugo.io/t/sorting-menu-by-weight-and-reversed/27710) for more information.

### To reproduce the problem

```bash
git clone https://github.com/pointyfar/hugo-menu-issue-27710
cd hugo-menu-issue-27710/
./test.sh
```

The test script will repeatedly build the site, compare /public/index.html to the baseline stored in the repository, and stop upon failure or after 200 iterations.

For me, the failure typically occurs after < 10 iterations, though I've seen it run fine for ~90 iterations, then fail.

**Comment by satotake** [comment]
@jmooring 

I'm running into this issue.
Your sample and script are very useful for debugging, thank you.

Could I make sure one thing?
In my opinion, there are not only a race but also another issue.

In [sample template](https://github.com/pointyfar/hugo-menu-issue-27710/blob/master/layouts/_default/baseof.html),


```html
    {{ range $k, $v := .Site.Menus.main }}
    {{ $k }} / {{ $v.Weight }} / {{ $v.Name }} / <a href="{{.URL}}">{{ $v.Page }}</a>
    <br>
    {{ end }}
```


First range is expected to be rendered like so

```html
    0 / 100 / Home / <a href="/">Page(/_index.md)</a>
    <br>

    1 / 50 / E / <a href="/posts/e/">Page(/posts/E.md)</a>
    <br>
    
    2 / 40 / D / <a href="/posts/d/">Page(/posts/D.md)</a>
    <br>
    
    3 / 30 / C / <a href="/posts/c/">Page(/posts/C.md)</a>
    <br>
    
    4 / 20 / B / <a href="/posts/b/">Page(/posts/B.md)</a>
    <br>
    
    5 / 10 / A / <a href="/posts/a/">Page(/posts/A.md)</a>
    <br>
    
    <hr>
```

In my opinion, it is strange 

- default sort is used here but the order is same as `Reverse`
- if we remove `Reverse` range from [sample template](https://github.com/pointyfar/hugo-menu-issue-27710/blob/master/layouts/_default/baseof.html),
current Hugo renders default sort like `ByWeight`

Is it intentional behaviour?

**Comment by jmooring** [comment]
@satotake I don't understand your question. I pulled #8561 locally and tested.

```toml
[[menu.main]]
name = "a"
url = "/a/"
weight = 20

[[menu.main]]
name = "b"
url = "/b/"
weight = 10

[[menu.main]]
name = "c"
url = "/c/"
weight = 30
```

Results:

```text
range $k, $v := site.Menus.main

b (weight 10)
a (weight 20)
c (weight 30)

range $k, $v := site.Menus.main.Reverse

c (weight 30)
a (weight 20)
b (weight 10)

range $k, $v := site.Menus.main.ByWeight

b (weight 10)
a (weight 20)
c (weight 30)

range $k, $v := site.Menus.main.ByWeight.Reverse

c (weight 30)
a (weight 20)
b (weight 10)

range $k, $v := site.Menus.main.ByName

a (weight 20)
b (weight 10)
c (weight 30)

range $k, $v := site.Menus.main.ByName.Reverse

c (weight 30)
b (weight 10)
a (weight 20)
```

This looks right to me.

**Comment by satotake** [comment]
@jmooring 
I am wondering if .Site.Menus should be mutable or immutable in terms of specification.
Before #8561, the order can be mutated
After #8561 , the order is always same.

> This looks right to me.

If so, it will be okay.
Thanks for checking.

**Comment by satotake** [comment]
p.s.
If #8561  is right, we need to modify test repo like [this](https://github.com/satotake/hugo-menu-issue-27710/commit/b10c569f53905761275de2814620b9456cfeb03d) to pass your test.
